### PR TITLE
multicluster: Create sni_cluster filter for the Multicluster Gateway listener

### PR DESCRIPTION
**Pull request by draychev**
This PR configures a `sni_cluster` filter for the Multicluster Gateway listener.

### Example of this looks like

> Note `envoy.filters.network.sni_cluster`

```json
  {
    "@type": "type.googleapis.com/envoy.admin.v3.ListenersConfigDump",
    "version_info": "1",
    "dynamic_listeners": [
      {
        "name": "multicluster-listener",
        "active_state": {
          "version_info": "1",
          "listener": {
            "@type": "type.googleapis.com/envoy.config.listener.v3.Listener",
            "name": "multicluster-listener",
            "address": {
              "socket_address": {
                "address": "0.0.0.0",
                "port_value": 15003
              }
            },
            "filter_chains": [
              {
                "filter_chain_match": {
                  "transport_protocol": "tls",
                  "application_protocols": [
                    "osm"
                  ],
                  "server_names": [
                    "*.local"
                  ]
                },
                "filters": [
                  {
                    "name": "envoy.filters.network.sni_cluster"
                  },
                  {
                    "name": "envoy.filters.network.tcp_proxy",
                    "typed_config": {
                      "@type": "type.googleapis.com/envoy.extensions.filters.network.tcp_proxy.v3.TcpProxy",
                      "stat_prefix": "passthrough-multicluster-gateway",
                      "cluster": "passthrough-multicluster-gateway",
                      "access_log": [
                        {
                          "name": "envoy.access_loggers.stream",
                          "typed_config": {
...
```
Signed-off-by: Delyan Raychev <delyan.raychev@microsoft.com>

### CODE REVIEW

**Review by snehachhabria — APPROVED** [review]

File: pkg/envoy/lds/gateway.go [hunk 1]
```diff
@@ -1,90 +1,89 @@
...
+	inboundMulticlusterGatewayFilterChainName = "inbound-multicluster-gateway-filter-chain"
 )
 
 func (lb *listenerBuilder) buildGatewayListeners() []types.Resource {
 	if !lb.cfg.GetFeatureFlags().EnableMulticlusterMode {
 		return nil
 	}
+
+	// TODO(draychev): what should this be?
+	svcIdent := identity.ServiceIdentity("osm-gateway.osm-system")
```
Comment (shashankram) [inline pkg/envoy/lds/gateway.go:26]: Why is this required? This should already be set as a part of the `listenerBuilder` struct.

File: pkg/envoy/xdsutil.go [hunk 2]
```diff
@@ -407,3 +411,12 @@ func GetServiceIdentityFromProxyCertificate(cn certificate.CommonName) (identity
...
 }
+
+// MessageToAny converts from proto message to proto Any and returns an error if any
+func MessageToAny(pb proto.Message) (*any.Any, error) {
+	msg, err := ptypes.MarshalAny(pb)
+	if err != nil {
+		return nil, err
+	}
+	return msg, nil
+}
```
Comment (shashankram) [inline pkg/envoy/xdsutil.go:422]: Let's remove this and use `ptypes.MarshalAny` directly.

**Review by shashankram — CHANGES_REQUESTED** [review]
2 comments, looks good otherwise: https://github.com/openservicemesh/osm/pull/3812/files#r673550702
https://github.com/openservicemesh/osm/pull/3812/files#r673551069

File: pkg/envoy/lds/gateway.go (hunk 1 above)
Comment (draychev) [reply pkg/envoy/lds/gateway.go]: I'm patching this and handing it off to @snehachhabria.  I can't remember why we need it here (bringing changes from a branch from December 2020).

**Review by shashankram — APPROVED** [review]

File: pkg/envoy/lds/gateway.go (hunk 1 above)
Comment (draychev) [reply pkg/envoy/lds/gateway.go]: Per @shashankram - used https://github.com/openservicemesh/osm/blob/main/cmd/osm-controller/gateway.go#L44
